format/checkpoint_test.go

```diff
@@ -0,0 +1,183 @@
+			want: format.Checkpoint{
+				Ecosystem: "Log Checkpoint v0",
+				Size:      123,
+				RootHash:  []byte("bananas"),
+			},
+		}, {
+			desc: "valid with different ecosystem",
+			m:    "Banana Checkpoint v1\n9944\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\n",
+			want: format.Checkpoint{
+				Ecosystem: "Banana Checkpoint v1",
+				Size:      9944,
+				RootHash:  []byte("the view from the tree tops is great!"),
+			},
+		}, {
+			desc: "valid with trailing data",
+			m:    "Log Checkpoint v0\n9944\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\nHere's some associated data.\n",
+			want: format.Checkpoint{
+				Ecosystem: "Log Checkpoint v0",
+				Size:      9944,
+				RootHash:  []byte("the view from the tree tops is great!"),
+			},
+			wantRest: []byte("Here's some associated data.\n"),
+		}, {
+			desc:    "invalid empty header",
+			m:       "\n9944\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\n",
+			wantErr: true,
+		}, {
+			desc:    "invalid size - not a number",
+			m:       "Log Checkpoint v0\nbananas\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\n",
+			wantErr: true,
+		}, {
+			desc:    "invalid size - negative",
+			m:       "Log Checkpoint v0\n-34\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\n",
+			wantErr: true,
+		}, {
+			desc:    "invalid size - too large",
+			m:       "Log Checkpoint v0\n3438945738945739845734895735\ndGhlIHZpZXcgZnJvbSB0aGUgdHJlZSB0b3BzIGlzIGdyZWF0IQ==\n",
+			wantErr: true,
+		}, {
+			desc:    "invalid roothash - not base64",
+			m:       "Log Checkpoint v0\n123\nThisIsn'tBase64\n",
+			wantErr: true,
+		},
+	} {
+		t.Run(string(test.desc), func(t *testing.T) {
+			var got format.Checkpoint
+			var gotErr error
+			var gotRest []byte
+			if gotRest, gotErr = got.Unmarshal([]byte(test.m)); (gotErr != nil) != test.wantErr {
+				t.Fatalf("Unmarshal = %q, wantErr: %T", gotErr, test.wantErr)
+			}
+			if diff := cmp.Diff(test.want, got); len(diff) != 0 {
+				t.Fatalf("Unmarshalled Checkpoint with diff %s", diff)
+			}
+			if !bytes.Equal(test.wantRest, gotRest) {
+				t.Fatalf("got rest %x, want %x", gotRest, test.wantRest)
+			}
+		})
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Below is an example of embedding the minimal checkpoint as one way to extend
+// it to include additional ecosystem-specific data.
+// Reimplementing parsing of the full extended structure would be fine too.
+
+// moonLogCheckpoint is a hypothetical checkpoint for an ecosystem which requires
+// its checkpoints to commit to more data than the minimum common checkpoint does.
+type moonLogCheckpoint struct {
+	format.Checkpoint
+	Timestamp uint64
+	Phase     string
+}
+
+// Unmarshal knows how to unmarshal the moon log data.
+// It delegates to the embedded Checkpoint to unmarshal itself first, before
+// attempting to unmarshal the Moon ecosystem specific data.
+func (m *moonLogCheckpoint) Unmarshal(data []byte) error {
```

Comment:
Maybe also add a Marshal method to this to show how the extra data would be poked on to the end of the string. For completeness. Or leave it as an exercise to the reader.

Reply:
Added (along with a roundtrip test)